sklearn/_fast_gradient_boosting/grower.py

```diff
@@ -0,0 +1,481 @@
+            self._finalize_leaf(left_child_node)
+        if right_child_node.n_samples < self.min_samples_leaf * 2:
+            self._finalize_leaf(right_child_node)
+
+        # Compute histograms of childs, and compute their best possible split
+        # (if needed)
+        should_split_left = left_child_node.value is None  # node isn't a leaf
+        should_split_right = right_child_node.value is None
+        if should_split_left or should_split_right:
+
+            # We will compute the histograms of both nodes even if one of them
+            # is a leaf, since computing the second histogram is very cheap
+            # (using histogram subtraction).
+            n_samples_left = left_child_node.sample_indices.shape[0]
+            n_samples_right = right_child_node.sample_indices.shape[0]
+            if n_samples_left < n_samples_right:
+                smallest_child = left_child_node
+                largest_child = right_child_node
+            else:
+                smallest_child = right_child_node
+                largest_child = left_child_node
+
+            # We use the brute O(n_samples) method on the child that has the
+            # smallest number of samples, and the subtraction trick O(n_bins)
+            # on the other one.
+            tic = time()
+            smallest_child.histograms = \
+                self.histogram_builder.compute_histograms_brute(
+                    smallest_child.sample_indices)
+            largest_child.histograms = \
+                self.histogram_builder.compute_histograms_subtraction(
+                    node.histograms, smallest_child.histograms)
+            self.total_compute_hist_time += time() - tic
+
+            tic = time()
+            if should_split_left:
+                self._compute_best_split_and_push(left_child_node)
+            if should_split_right:
+                self._compute_best_split_and_push(right_child_node)
+            self.total_find_split_time += time() - tic
+
+        return left_child_node, right_child_node
+
+    def can_split_further(self):
+        """Return True if there are still nodes to split."""
+        return len(self.splittable_nodes) >= 1
+
+    def _finalize_leaf(self, node):
+        """Compute the prediction value that minimizes the objective function.
+
+        This sets the node.value attribute (node is a leaf iff node.value is
+        not None).
+
+        See Equation 5 of:
+        XGBoost: A Scalable Tree Boosting System, T. Chen, C. Guestrin, 2016
+        https://arxiv.org/abs/1603.02754
+        """
+        node.value = -self.shrinkage * node.sum_gradients / (
+            node.sum_hessians + self.splitter.l2_regularization)
+        self.finalized_leaves.append(node)
+
+    def _finalize_splittable_nodes(self):
+        """Transform all splittable nodes into leaves.
+
+        Used when some constraint is met e.g. maximum number of leaves or
+        maximum depth."""
+        while len(self.splittable_nodes) > 0:
```

Comment:
I guess it would be very marginally faster, but I like the idea  that `splittable_nodes` ends up being empty at the end, and that the tree keeps being built in a best-first-search fashion.

No strong feeling though